quantile/tdigest/tdigest.go

```diff
@@ -0,0 +1,285 @@
+		currUnmerged := sentinelCentroid
+		if len(unmerged) > 0 {
+			currUnmerged = unmerged[0]
+		}
+
+		currMerged := sentinelCentroid
+		// buffer takes precedence over merged
+		if len(buffer) > 0 {
+			currMerged = buffer[0]
+		} else if len(merged) > 0 {
+			currMerged = merged[0]
+		}
+
+		if currUnmerged.Mean < currMerged.Mean {
+			// Save the first centroid in case it'll be overwritten
+			if len(merged) > 0 {
+				buffer = d.appendCentroid(buffer, merged[0])
+				merged = merged[1:]
+			}
+			unmerged = unmerged[1:]
+			currIndex = d.mergeCentroid(currIndex, currWeight, totalWeight, currUnmerged)
+			currWeight += currUnmerged.Weight
+		} else {
+			if len(merged) > 0 {
+				// Save the first centroid in case it'll be overwritten
+				if len(buffer) > 0 {
+					copy(buffer, buffer[1:])
+					buffer[len(buffer)-1] = merged[0]
+				}
+				merged = merged[1:]
+			} else {
+				// Merged buffer is used up, now consuming saved centroids
+				buffer = buffer[1:]
+			}
+			currIndex = d.mergeCentroid(currIndex, currWeight, totalWeight, currMerged)
+			currWeight += currMerged.Weight
+		}
+	}
+
+	d.unmerged = d.unmerged[:0]
+	d.unmergedWeight = 0.0
+	d.mergedWeight = totalWeight
+}
+
+// mergeCentroid merges a centroid into the list of merged centroids
+func (d *tDigest) mergeCentroid(
+	currIndex float64,
+	currWeight float64,
+	totalWeight float64,
+	c Centroid,
+) float64 {
+	nextIndex := d.nextIndex((currWeight + c.Weight) / totalWeight)
+	if nextIndex-currIndex > 1 || len(d.merged) == 0 {
+		// This is the first centroid added, or the next index is too far away from the current index
+		d.merged = d.appendCentroid(d.merged, c)
+		return d.nextIndex(currWeight / totalWeight)
+	}
+
+	// The new centroid falls within the range of the current centroid
+	numMerged := len(d.merged)
+	d.merged[numMerged-1].Weight += c.Weight
+	d.merged[numMerged-1].Mean += (c.Mean - d.merged[numMerged-1].Mean) * c.Weight / d.merged[numMerged-1].Weight
+	return currIndex
+}
+
+// nextIndex estimates the index of the next centroid
+func (d *tDigest) nextIndex(quantile float64) float64 {
+	return d.compression * (math.Asin(2*quantile-1)/math.Pi + 0.5)
+}
+
+// add adds a weighted value
+func (d *tDigest) add(value float64, weight float64) {
+	if len(d.unmerged) == d.unmergedCapacity {
+		d.compress()
+	}
+	d.minValue = math.Min(d.minValue, value)
+	d.maxValue = math.Max(d.maxValue, value)
+	d.unmerged = d.appendCentroid(d.unmerged, Centroid{Mean: value, Weight: weight})
+	d.unmergedWeight += weight
+}
+
+// upperBound returns the upper bound for computing quantiles given the centroid index.
+// d.merged is guaranteed to have at least one centroid when upperBound is called.
+func (d *tDigest) upperBound(index int) float64 {
+	if index == len(d.merged)-1 {
+		return d.maxValue
+	}
+	return (d.merged[index].Mean + d.merged[index+1].Mean) / 2.0
+}
+
+func (d *tDigest) appendCentroid(centroids []Centroid, c Centroid) []Centroid {
+	if len(centroids) == cap(centroids) {
```

Comment:
From `compress`, it only resets `buffer` pointer when `len(merged) == 0`, and once that is true, we'll never call `appendCentroid` afterwards, so this should be a non issue.